### PR TITLE
refactor(react-components): Make generic hook for handling a property on any command

### PR DESCRIPTION
**Pull request by nilscognite**
#### Type of change
Refactor 

## Description :pencil:

3 new hooks, `useCommand ` , `useCommandProperty` and `useCommandProps`. **These are made the code easier to read and understand, by removing repetitive and redundant code lines.**

Also made some one liner hooks for each specific property, like `useCommandIcon`, `useCommandUniqueId`, `useCommandVisible`, `useCommandEnable `, `useCommandChecked `and `useSliderCommandValue `for convenience.

I have rolled it out in all react component that is under unit test. It should not change anything, this is just a simple refactoring.
The rest is taken later. To find them, search for `useOnUpdate`.. This method should be removed when finished. 

During this work I have made the following new unit test:

* SegmentedButtons
* useGhostMode

**This is a first iteration to make the command based component easier to read and use.**

Since the coverage went down I have another PR that test `BaseFilterCommand`.

### CODE REVIEW

File: react-components/src/hooks/useGhostMode.test.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+describe(useGhostMode.name, () => {
+  let renderTarget: RevealRenderTarget;
+  let command: SetGhostModeCommand;
+
+  beforeAll(() => {
+    renderTarget = createFullRenderTargetMock();
+    // This add the SetGhostModeCommand to the list of available commends
+    command = getDefaultCommand(new SetGhostModeCommand(), renderTarget);
+  });
```
Comment (haakonflatval-cognite) [inline react-components/src/hooks/useGhostMode.test.tsx]: Should this be `beforeEach`?

File: react-components/src/components/Architecture/SegmentedButtons.test.tsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    expect(afterButtons.length).toBe(2);
+    expect(afterButtons[0].getAttribute('aria-selected')).toBe('false');
+    expect(afterButtons[1].getAttribute('aria-selected')).toBe('true');
+
+    expect(command.children[0].isChecked).toBe(false);
+    expect(command.children[1].isChecked).toBe(true);
+  });
+});
+
+function renderMe(command: BaseOptionCommand): void {
```
Comment (haakonflatval-cognite) [inline react-components/src/components/Architecture/SegmentedButtons.test.tsx]: Could we have a more descriptive name for this? :sweat_smile:

**Review by haakonflatval-cognite — DISMISSED** [review]
A couple of nitpicks, but nothing mandatory :+1: I'll approve again if you fix it

File: react-components/src/components/Architecture/SegmentedButtons.test.tsx (hunk 2 above)
Comment (nilscognite) [reply react-components/src/components/Architecture/SegmentedButtons.test.tsx]: Fix: deba84b6ea96b6892ae3b83b08aba43253246205

File: react-components/src/hooks/useGhostMode.test.tsx (hunk 1 above)
Comment (nilscognite) [reply react-components/src/hooks/useGhostMode.test.tsx]: Fix: deba84b6ea96b6892ae3b83b08aba43253246205

**Review by haakonflatval-cognite — APPROVED** [review]
LGTM :+1:

**Review by haakonflatval-cognite — APPROVED** [review]
LGTM again :+1:

File: react-components/src/components/Architecture/SegmentedButtons.test.tsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+import { act, type PropsWithChildren, type ReactElement } from 'react';
+import { viewerMock } from '#test-utils/fixtures/viewer';
+import { sdkMock } from '#test-utils/fixtures/sdk';
+import { ViewerContextProvider } from '../RevealCanvas/ViewerContext';
+import { CommandButton } from './CommandButton';
+import { OptionType } from '../../architecture/base/commands/BaseOptionCommand';
+import { SegmentedButtons } from './SegmentedButtons';
+import userEvent from '@testing-library/user-event';
+
+// Help page here:  https://bogr.dev/blog/react-testing-intro/
```
Comment (pramod-cog) [inline react-components/src/components/Architecture/SegmentedButtons.test.tsx]: remove the comments

File: react-components/src/components/Architecture/SegmentedButtons.test.tsx [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+const TEST_ID = 'segmented-control-button';
+
+describe(CommandButton.name, () => {
+  afterEach(() => {
+    cleanup();
+  });
+
+  test('should render with default values', async () => {
```
Comment (pramod-cog) [inline react-components/src/components/Architecture/SegmentedButtons.test.tsx]: (suggestion)
async not needed here

File: react-components/src/components/Architecture/SegmentedButtons.test.tsx [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+      const button = buttons[i];
+      const option = command.children[i];
+
+      expect(button.getAttribute('type')).toBe('button');
+      expect(button.getAttribute('aria-disabled')).toBe('false');
+      expect(button.getAttribute('aria-label')).toBe(option.label);
+      expect(button.getAttribute('aria-selected')).toBe(option.isChecked.toString());
+
+      const buttonClass = button.getAttribute('class');
+      expect(buttonClass).not.contains('toggled');
```
Comment (pramod-cog) [inline react-components/src/components/Architecture/SegmentedButtons.test.tsx]:
```suggestion
      expect(buttonClass).not.toContain('toggled');
```

File: react-components/src/components/Architecture/SegmentedButtons.test.tsx [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    const command = new MockSegmentedCommand();
+    renderSegmentedButtons(command);
+
+    const beforeButtons = screen.queryAllByTestId(TEST_ID);
+    expect(beforeButtons.length).toBe(2);
+
+    act(() => {
+      command.isVisible = false;
+    });
+    const afterButtons = screen.queryAllByTestId(TEST_ID);
```
Comment (pramod-cog) [inline react-components/src/components/Architecture/SegmentedButtons.test.tsx:55]: (clarification)
Using `act()` is good, however after changing `.isVisible`, you query the DOM, if update is `async` then to ensure DOM updates before assertions waitFor should be considered.

File: react-components/src/components/Architecture/SegmentedButtons.test.tsx (hunk 5 above)
Comment (pramod-cog) [reply react-components/src/components/Architecture/SegmentedButtons.test.tsx]: Current code may not work as intended.

File: react-components/src/components/Architecture/SegmentedButtons.test.tsx [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+import { SegmentedButtons } from './SegmentedButtons';
+import userEvent from '@testing-library/user-event';
+
+// Help page here:  https://bogr.dev/blog/react-testing-intro/
+
+const TEST_ID = 'segmented-control-button';
+
+describe(CommandButton.name, () => {
+  afterEach(() => {
+    cleanup();
```
Comment (pramod-cog) [inline react-components/src/components/Architecture/SegmentedButtons.test.tsx]: `cleanup` is triggered automatically by the `testing-library`

File: react-components/src/components/Architecture/hooks/useCommandProperty.tsx [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import { useState, useEffect, useCallback } from 'react';
+import { type BaseCommand } from '../../../architecture/base/commands/BaseCommand';
+
+export function useCommandProperty<T>(command: BaseCommand, getValue: () => T): T {
+  const [value, setValue] = useState<T>(getValue);
```
Comment (pramod-cog) [inline react-components/src/components/Architecture/hooks/useCommandProperty.tsx:5]: (suggestion)
Setting initial value is redundant, `useEffect` gets called which triggers `memoizedUpdate`, but this is no harm 😄

File: react-components/src/hooks/useGhostMode.test.tsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import { beforeEach, describe, expect, test } from 'vitest';
+
+import { useGhostMode } from './useGhostMode';
+import { type RevealRenderTarget } from '../architecture';
+import { ViewerContextProvider } from '../components/RevealCanvas/ViewerContext';
+import { createFullRenderTargetMock } from '../../tests/tests-utilities/fixtures/createFullRenderTargetMock';
+import { SetGhostModeCommand } from '../architecture/base/concreteCommands/cad/SetGhostModeCommand';
+import { getDefaultCommand } from '../components/Architecture/utilities';
+
+describe(useGhostMode.name, () => {
```
Comment (pramod-cog) [inline react-components/src/hooks/useGhostMode.test.tsx:10]: (suggestion)
Can we add test for edge cases where `command` is missing and `content` is not provided

File: react-components/src/components/Architecture/SegmentedButtons.test.tsx (hunk 6 above)
Comment (nilscognite) [reply react-components/src/components/Architecture/SegmentedButtons.test.tsx:55]: Not async

File: react-components/src/components/Architecture/hooks/useCommandProperty.tsx (hunk 8 above)
Comment (nilscognite) [reply react-components/src/components/Architecture/hooks/useCommandProperty.tsx:5]: I have to set something. Otherwise it will be undefined.

File: react-components/src/components/Architecture/SegmentedButtons.test.tsx (hunk 7 above)
Comment (nilscognite) [reply react-components/src/components/Architecture/SegmentedButtons.test.tsx]: Removed

File: react-components/src/hooks/useGhostMode.test.tsx (hunk 9 above)
Comment (nilscognite) [reply react-components/src/hooks/useGhostMode.test.tsx:10]: Made a new test in useGhostMode.test

**Review by pramod-cog — COMMENTED** [review]
Looks good from risk reviewer intern point of view

**Comment by nilscognite** [comment]
Changes from prev review are here 2498de40d88010b77d8bc8a03f0780c8365e28df

**Comment by pramod-cog** [comment]
🦄

**Review by pramod-cog — APPROVED** [review]

**Review by haakonflatval-cognite — APPROVED** [review]
🤞
